### ISSUE TITLE
Unable to run tests

### ISSUE THREAD

**Issue by michas2** [tass-belgium/picotcp#434]
I'm new to picotcp and tried to [build](https://github.com/tass-belgium/picotcp/wiki/Setting-up-the-environment) picotcp and [run](https://github.com/tass-belgium/picotcp/wiki/Testing) the tests.

For this I used the following Dockerfile:
```
FROM ubuntu

RUN apt-get update && apt-get -y install git check vde2 libvdeplug2-dev libpcap0.8-dev openvpn make gcc

RUN git clone -b v1.6.2 https://github.com/tass-belgium/picotcp.git

WORKDIR /picotcp

RUN make
RUN make clean; make units ARCH=faulty && ./test/units.sh
```

The build works fine, but the tests seem not to build correctly.
There are lots of warnings starting here:

```
	[UNIT TESTS SUITE]
-e 	[CC] units.o
In file included from test/units.c:25:0:
modules/pico_tcp.c:46:0: warning: "IS_NAGLE_ENABLED" redefined
 #define IS_NAGLE_ENABLED(s)     (!(!(!(s->opt_flags & (1u << PICO_SOCKET_OPT_TCPNODELAY)))))
[...]
In file included from test/units.c:59:0:
test/unit/unit_mocks.c:3:5: warning: no previous prototype for 'mock_print_protocol' [-Wmissing-prototypes]
 int mock_print_protocol(uint8_t *buf)
     ^
[...]
In file included from test/units.c:62:0:
test/unit/unit_dhcp.c:4:6: warning: no previous prototype for 'callback_dhcpclient' [-Wmissing-prototypes]
 void callback_dhcpclient(void*cli, int code)
[...]
```
And a final error:
```
-e 	[LD] /picotcp/build/test/units
/usr/lib/gcc/x86_64-linux-gnu/5/../../../x86_64-linux-gnu/libcheck.a(check_log.o): In function `subunit_lfun':
(.text+0x584): undefined reference to `subunit_test_start'
/usr/lib/gcc/x86_64-linux-gnu/5/../../../x86_64-linux-gnu/libcheck.a(check_log.o): In function `subunit_lfun':
(.text+0x63f): undefined reference to `subunit_test_fail'
/usr/lib/gcc/x86_64-linux-gnu/5/../../../x86_64-linux-gnu/libcheck.a(check_log.o): In function `subunit_lfun':
(.text+0x654): undefined reference to `subunit_test_pass'
/usr/lib/gcc/x86_64-linux-gnu/5/../../../x86_64-linux-gnu/libcheck.a(check_log.o): In function `subunit_lfun':
(.text+0x66f): undefined reference to `subunit_test_error'
collect2: error: ld returned 1 exit status
make: *** [units] Error 1
```

For a full output please build the above Dockerfile.
Any idea what is wrong here?

---

Also, is there any chance to run the tests on alpine linux?
I used exactly the same commands as above, only installed packages using
`apk add make git vde2 openvpn check check-dev libpcap-dev bash gcc musl-dev coreutils`.

The same warnings, but the error is about missing libasan. (which seems not to be directly available on alpine linux.)
Is there a good way to run test here. (probably by deactivating/ignoring libasan.)

**Comment by oussemah** [comment]
Hello @michas2 

Can you  please check that you have subunit package installed ?
It's needed for tests.

**Comment by frederikvs** [comment]
For the warnings in building the unit tests : on everything that's been released up till now, warnings in building the unit tests were ignored. Since then we've worked on it, and on the current development branch (which we hope to release soonish), the unit tests should also be warningfree. But for now, just ignore those. 
The errors you see are, as @oussemah pointed out, probably because of a dependency on the `subunit` package. A recent version of LibCheck (included in ubuntu 16.04, a.o.) silently added a dependency on libsubunit, so now you need to have libsubunit installed, and you need to add `-lsubunit` to the linker flags. You could cherry-pick commit https://github.com/tass-belgium/picotcp/commit/ac8462d9b67115e80ec88ca42f8723a3cb116897 for this, or just add it manually.

As for the address sanitizer problems : could you try running with `ADDRESS_SANITIZER=0` added to your make command?

**Comment by michas2** [comment]
Thanks a lot for the quick answer!
I first tested with the master branch and the (latest) v1.6.2 tags and both suffered from the described subunit issue.

After switching to the development branch (which contains ac8462d) and using `UNITS=1` the tests ran fine. :)

It now also works fine on alpine. Adding `ADDRESS_SANITIZER=0` and explicitly installing  the `subunit-dev` package did the trick there. :)
